### ISSUE TITLE
useEventBus causes memory leaks in some scenarios

### ISSUE THREAD

**Issue by Renhor** [vueuse/vueuse#3498]
### Describe the bug

Combination with pinia storeToRefs + SSR (nuxt3) + useEventBus().on causes memory leaks.

I'm not sure it directly related to useEventBust or storeToRefs pinia or the way i use them together ;(

index.vue
```
const store = useSomeStore();
const { form } = storeToRefs(store);

// ok
// useSomeComponentLogic(store.form);

// cause leaks
useSomeComponentLogic(form);
```

useSomeComponentLogic.ts
```
export const useSomeComponentLogic = (_ref: MaybeRef<{ data: string }>) => {
  const bus = useAppEventBus(AppEvent.Test);
  const ref = computed(() => unref(_ref));

  // causes leak

  bus.on(() => {
    ref.value.data = 'test';
    console.log('im bus');
  });

  // ok

  // onMounted(() => {
  //   bus.on(() => {
  //     ref.value.data = 'test';
  //     console.log('im bus');
  //   });
  // })
};
```

### Reproduction

https://github.com/Renhor/useEventBus-leak

### System Info

```Shell
System:
    OS: macOS 14.0
    CPU: (10) arm64 Apple M1 Max
    Memory: 6.37 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.17.1 - ~/.nvm/versions/node/v18.17.1/bin/node
    npm: 9.6.7 - ~/.nvm/versions/node/v18.17.1/bin/npm
  Browsers:
    Chrome: 118.0.5993.88
    Safari: 17.0
  npmPackages:
    @vueuse/core: ^10.5.0 => 10.5.0 
    @vueuse/nuxt: ^10.5.0 => 10.5.0 
    vue: ^3.3.6 => 3.3.6
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by Renhor** [comment]
Up

**Comment by datWeazel** [comment]
I spent the last week trying to find a memory leak in my Nuxt3 projects and the event listener "issue" was the root cause of all of them. 

Our team trusted that the following detail in the useEventBus documentation would also be true for Nuxt3 codebases: 
`Listeners registered inside of components setup will be unregistered automatically when the component gets unmounted.`
https://vueuse.org/core/useEventBus/

But as it seems that either isn't the case or we did something else wrong. 

Long story short. 
Changing
```ts
<script setup lang="ts">
const myEventBus = useEventBus<boolean>(true);
myEventBus.on(() => {
  // do something
});
</script>
```

to 
```ts
<script setup lang="ts">
function test(): void { // do something };

onMounted(() => {
  myEventBus.on(test);
});

onBeforeUnmount(() => {
  myEventBus.off(test);
});
</script>
```

Fixed our memory leak.

**Comment by Piscinelove** [comment]
Thanks @datWeazel !
I spent 2 days looking for the cause of our memory leaks until I came across your comment. The memory leak did come from using useEventBus. The problem is also present when using Vue 2.7.

**Comment by MosheL** [comment]
I think useEventBus need to sent a empty mockup on SSR instance.  just ignore the on().

and for now

    if(process.browser) eventBus.on(.....

**Comment by Jesus82** [comment]
This should be either documented or fixed, it really breaks an app in production, and it is almost impossible to debug...

**Comment by JianGuoHt** [comment]
The document says：
Listeners registered inside of components setup will be unregistered automatically when the component gets unmounted.

But that's not how it works in practice

**Comment by samerdraidi** [comment]
I can confirm that there is an issue. the IDE ts server breaks once I start using it specially inside pinia store. even when SSR false

**Comment by jackcodetw** [comment]
@samerdraidi how do you use it, do you have any minimal reproduction or examples?

**Comment by th3l0g4n** [comment]
I also have no reproduction, but I can confirm that this is indeed an issue. Upon reloading our app, the memory consumption skyrockets and the mem is never released.

The eventbus is part of a pinia store and my solution in the end was to export it like this:
```ts
return {
  onEvent: import.meta.browser ? emitter.on : () => () => {},
}
```
